subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,542 @@
+					     &response.image_upload_offset),
+		ZCBOR_MAP_DECODE_KEY_DECODER("rc", zcbor_int32_decode, &response.status)};
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+		upload_state.upload_status = MGMT_ERR_ETIMEOUT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return 0;
+	}
+
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	rc = zcbor_map_decode_bulk(zsd, upload_res_decode, ARRAY_SIZE(upload_res_decode), &decoded);
+	if (rc || response.status == -1 || response.image_upload_offset == SIZE_MAX) {
+		response.status = MGMT_ERR_ECORRUPT;
+		upload_state.upload_status = MGMT_ERR_ECORRUPT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return MGMT_ERR_EINVAL;
+	}
+
+	upload_state.upload_status = response.status;
+	upload_state.offset = response.image_upload_offset;
+	if (upload_state.request_length == upload_state.writed_length) {
+		client_gr_image_fn(&response);
+	}
+
+	k_sem_give(user_data);
+	return 0;
+}
+
+static int erase_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_image_gr_rsp response;
+	zcbor_state_t zsd[CBOR_ENCODER_STATE_NUM + 2];
+	size_t decoded;
+	int rc;
+
+	response.status = -1;
+
+	struct zcbor_map_decode_key_val upload_res_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("rc", zcbor_int32_decode, &response.status)};
+
+	response.cmd = IMG_MGMT_ID_ERASE;
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return 0;
+	}
+
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	rc = zcbor_map_decode_bulk(zsd, upload_res_decode, ARRAY_SIZE(upload_res_decode), &decoded);
+	if (rc || response.status == -1) {
+		LOG_ERR("Erase fail %d", rc);
+		response.status = MGMT_ERR_ECORRUPT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return MGMT_ERR_EINVAL;
+	}
+
+	client_gr_image_fn(&response);
+	k_sem_give(user_data);
+	return 0;
+}
+
+int img_mgmt_client_callback_register(mgmt_client_gr_image_fn image_fn)
+{
+	client_gr_image_fn = image_fn;
+	return 0;
+}
+
+int img_mgmt_client_upload_init(size_t image_size, int image_num)
+{
+	if (!client_gr_image_fn) {
+		return -1;
+	}
+	upload_state.image_size = image_size;
+	upload_state.offset = 0;
+	upload_state.image_num = image_num;
+	return 0;
+}
+
+int img_mgmt_client_upload(const uint8_t *data, size_t length)
+{
+	struct net_buf *nb;
+	const uint8_t *write_ptr;
+	int rc;
+	bool ok;
+	size_t write_length;
+	zcbor_state_t zse[CBOR_ENCODER_STATE_NUM + 2];
+
+	if (!client_gr_image_fn) {
+		return -1;
+	}
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+
+	upload_state.request_length = length;
+	upload_state.writed_length = 0;
+	upload_state.upload_status = 0;
+
+	while (upload_state.request_length != upload_state.writed_length) {
+		write_ptr = data + upload_state.writed_length;
+		write_length = upload_state.request_length - upload_state.writed_length;
+		if (write_length > 256) {
+			write_length = 256;
```

Comment:
Why 256?

Reply:
That was starting point value. I need to change this based on net_buf size.